Add explicit timestamp support for Span start/end.

carlosalberto opened the pull request:
A few notes:

* Both `end(Timestamp)` and `setStartTimestamp(Timestamp)` will report an error in case the values are `null`. This was done to keep uniformity with the rest of the API (although internally we are handling this case).
* A new `TimestampConverter` is created for `Span`s with explicit start time, and it's based on the aforementioned start time value.
* Internally (SDK) I'm creating/passing around `io.grpc.protobuf.Timestamp` instances for the explicit values, if any. `TimestampConverter` already requires one, anyway. Also, we get 'free' validation of the seconds/nanos.
  * HOWEVER, I'm not doing start/end timestamps comparison nor validation.

Fixes #24

carlosalberto commented:
Also fixes #545

jkwatson commented:
I think this is a good change. At some point in the future, it might be nice to not have the TimestampConverter us protobufs internally, but that can definitely wait.

adriancole commented:
one hazard is allowing SDK be a dependency kitchen sink. I would recommend making strict dependency review a gate for any public release as it will cause problems for people who pull in OTel even if they dont use it.

bogdandrutu commented:
@adriancole thanks for the comment. The main idea is that only the API artifact will be used as dependency in third-party libraries, and we try hard to limit the dependencies there.

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/Span.java:
```diff
@@ -189,6 +190,17 @@
+   * @param endTimestamp the explicit end {@link Timestamp} for this {@code Span}.
+   * @since 0.1.0
+   */
+  void end(Timestamp endTimestamp);
```
To avoid breaking people in the future, should we go with an option pattern here? That will let us add more properties in the future without changing this.

Not 100% convinced but want to hear options.

carlosalberto replied:
Sure, going with an option pattern sounds good to me. What about `EndOptions`, similar to what OC has? ;)

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/Span.java:
```diff
@@ -421,6 +433,15 @@
     Builder setSpanKind(Span.Kind spanKind);
 
+    /**
+     * Sets an explicit start {@link Timestamp} for the newly created {@code Span}.
```
Null is not accepted from what I see, then we should explain to the users:
1. when to set this.
2. what to do when this is not needed.

carlosalberto replied:
See your point. Will add extra clarification on this.

bogdandrutu replied:
I really want to discourage the following use patterns:

```java
Timestamp now = Timestamp.fromMillis(System.currentMillis());
tracer.spanBuilder().setStartTime(now).startSpan();
```
OR 
```java
tracer.spanBuilder().setStartTime(Timestamp.fromMillis(System.currentMillis())).startSpan();
```

Can you please phrase this in a way to make sure we discourage the examples I gave to you?

jkwatson replied:
What is the reason for discouraging this usage?

bogdandrutu replied:
There are few reasons here:
1) Java until java9 has only millis resolution (even java.time.Instant has only millis until java9).
2) We do want to be able to calculate all the events (in all the spans in one trace) using nanoTime relative to only one TimeConvertor (which is read only for the first Span of this trace in this process) so we can guarantee ordering of events. When using an absolute timestamp like java Instant or system.currentTimeMillis these are not monotonic so events are not order correctly.

jkwatson replied:
Thanks! I suspected it was along these lines. Thanks for confirming.

bogdandrutu replied:
Probably same comment should be for all the methods that allow startTime.

carlosalberto replied:
Yes (I know that we will also be adding explicit start time for `Event`s)

adriancole commented:
cool I will raise another issue then as the hard dep on grpc-context in the
api package can easily leak implementation and/or pin deps

https://github.com/open-telemetry/opentelemetry-java/blob/50a2db42b257cb9618131bffc2fbc997785b7f38/api/src/main/java/io/opentelemetry/distributedcontext/package-info.java

probably to really be an api you should move grpc integration to another
jar as there are alternative and arguably better context integrations.
there shouldn't be a favorite that pins a dep to grpc in other words. i
will copy this text to the new issue

On Sat, Sep 28, 2019, 10:15 PM Bogdan Drutu <notifications@github.com>
wrote:

> @adriancole <https://github.com/adriancole> thanks for the comment. The
> main idea is that only the API artifact will be used as dependency in
> third-party libraries, and we try hard to limit the dependencies there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/open-telemetry/opentelemetry-java/pull/571?email_source=notifications&email_token=AAAPVV3VICGR7P2DLGL4XILQL5RIBA5CNFSM4I3BNMFKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD722T7A#issuecomment-536193532>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAPVV3UC75GCS6D2NV2DL3QL5RIBANCNFSM4I3BNMFA>
> .
>

carlosalberto commented:
Hey @bogdandrutu 

I've updated the PR to not actually touch the SDK logic (as you suggested discussing the details for that in another PR). I've added a `EndSpanOptions` class (similar to what OC has) to contain an optional, nullable end `Timestamp`.

Things of interest:

1. `Timestamp` is an actual class. At first I had it as an interface, so vendors could extend it, but it feels like an overkill for such a small container.
2. Felt like adding a convenience `create(Timestamp endTimestamp)` method, so users don't have to create a builder when only setting this value, but didn't include it for cleanliness reasons.

Let me know ;)

bogdandrutu commented on api/src/main/java/io/opentelemetry/trace/EndSpanOptions.java:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @since 0.1
+   */
+  public static EndSpanOptions getDefault() {
```
Not sure this is needed for the public API. Let's remove until someone asks for it.

carlosalberto replied:
Works for me.

carlosalberto commented:
hey @bogdandrutu I've updated the code (making `EndSpanOptions.getDefault()` private). Let me know if this is good for you to merge (as discussed in the Java SIG last Friday).
